### ISSUE TITLE
Feature request: Permissions API

### ISSUE THREAD

**Issue by RendijsSmukulis** [MicrosoftEdge/WebView2Feedback#2427]
There's currently no way to list or manage site permissions (e.g. geolocation, camera, microphone) apart from handling the [PermissionsRequested](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2.permissionrequested?view=webview2-dotnet-1.0.1185.39) event.

To create meaningful indicators of permissions for the users we'd need to:
- Be able to list what permissions have been set for a given site (allow/deny/query)
- Be able to programmatically change and reset them

There's a couple ways this can be enabled:

- Quick-win: add a flag to WebView2 parameters that ensures the PermissionsRequested is fired every time a permission is requested by site. This would allow us to manage the permissions entirely in our code, and decide which ones are cached and when they're reset
- Alternatively, add a proper API that allows listing permissions that have been granted, denied, set to query; and modify/erase them.

[AB#39482825](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/39482825)

**Comment by victorthoang** [comment]
Thanks for the feedback RendijsSmukulis! We've added this as a scenario on our backlog!

**Comment by deeringc** [comment]
It's definitely not supported or even advisable to make any changes to these files, but it seems that the `EBWebView\Default\Preferences` json file stores the relevant information. 

~~~js
				"media_stream_camera": {
					"https://webcamtests.com:443,*": {
						"expiration": "0",
						"last_modified": "13298647470739859",
						"model": 0,
						"setting": 1
					}
				},
				"media_stream_mic": {},
~~~

**Comment by nishitha-burman** [comment]
Hi @RendijsSmukulis,

Can you describe the scenarios you are trying to light up by managing site permissions?

Thanks,
Nishitha

**Comment by RendijsSmukulis** [comment]
> Can you describe the scenarios you are trying to light up by managing site permissions?

- We want to provide the ability for our users to list what sites they have set non-default permissions on (e.g. always allow)
- We want to provide a way to modify or clear these permissions settings
- We want to provide a way to indicate to the user what permissions the user has granted to the currently open page, and modify said permissions

**Comment by nishitha-burman** [comment]
Hi @RendijsSmukulis,

Thank you for your reply. Other follow up questions: 
1. Do you want to programmatically reset permissions for a given site or control all permissions for all sites? Or both? 
2. Does this list cover all the permissions you are interested in: https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2permissionkind?view=webview2-dotnet-1.0.1185.39?

Thank you

**Comment by q71114** [comment]
Hi @nishitha-burman 

1. We want the former (to control permissions for any given site). That being said once that's done by monitoring which sites have permission set we can achieve the latter as well.
2. The most critical ones that we need to have access to are
   - Geolocation
   - Camera
   - Microphone
   - Screen Share <- this one is missing

**Comment by nishitha-burman** [comment]
Hi @q71114,

Do you want to be able to permissions for any site at any given time? E.g. be able to change permissions for a site even though you're not on that site? 

Thanks,
Nishitha

**Comment by q71114** [comment]
Hi @nishitha-burman,

Yes, we want the option to change the permission for any site at any given time although the scope for this is limited.

> E.g. be able to change permissions for a site even though you're not on that site?

We only want to do this when want to clear allowed / rejected permission so next time the permission request is raised again.

**Comment by nishitha-burman** [comment]
Hi @q71114,

Can you describe the end user scenario where you would do this?

Thank you,
Nishitha

**Comment by q71114** [comment]
> Can you describe the end user scenario where you would do this?

Hi @nishitha-burman,

We offer option to clear all data for specific domains. When selected all user data will be cleared for those domains including any permissions previously allowed / rejected by the user.

**Comment by nishitha-burman** [comment]
Hi @CoenraadS,

Is there an Edge (edge://settings/content) or web permission you are referring to that you would like to see triggered in this scenario?

**Comment by CoenraadS** [comment]
@nishitha-burman 

Actually I found another issue already for exactly this bug I mentioned: https://github.com/MicrosoftEdge/WebView2Feedback/issues/1039

**Comment by vickiez** [comment]
Hi all, we added some experimental permission management APIs in the new prerelease package, 1.0.1619-prerelease. 

- SavesInProfile property on [PermissionRequestedEventArgs](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2experimentalpermissionrequestedeventargs3?view=webview2-1.0.1619-prerelease)
- [SetPermissionState](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2experimentalprofile6?view=webview2-1.0.1619-prerelease#setpermissionstate) and [GetNonDefaultPermissionSettings](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2experimentalprofile6?view=webview2-1.0.1619-prerelease#getnondefaultpermissionsettings) on the profile
- New permission types: local font list, automatic downloads, media autoplay, file editing, system exclusive MIDI messages

You can read more here: https://github.com/MicrosoftEdge/WebView2Feedback/blob/main/specs/PermissionManagement.md. Please try these out and let us know if you run into any issues!

Steps for testing upcoming APIs/ features: https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/set-preview-channel.

**Comment by q71114** [comment]
Hi @vickiez, I can confirm that we can achieve what @RendijsSmukulis requested using the [SavesInProfile](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2experimentalpermissionrequestedeventargs3?view=webview2-1.0.1619-prerelease) property on PermissionRequestedEventArgs. Thanks for adding these APIs.

What is your timeline for having them in the stable release?

**Comment by vickiez** [comment]
Thanks for confirming @q71114. The APIs will be stable in next month's prerelease package, and then they will become part of the 111 release package that is tentatively scheduled for 3/9/23

**Comment by vickiez** [comment]
Hi all, the extended permission APIs are available in the latest release package, 1.0.1661.34. Closing this issue as completed but feel free to reactivate if needed. Thanks!
